### PR TITLE
Refactor aws_array_list harnesses with new proof style

### PR DESCRIPTION
**Pull request by feliperodri**
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

*Description of changes:*

This PR refactor the following proofs for `aws_array_list` with a new proof style:
 - aws_array_list_capacity
 - aws_array_list_clean_up
 - aws_array_list_clear
 - aws_array_list_copy

The refactor aims to improve readability and code quality of our proofs, which transit from an imperative style to a more functional one. This new proof style will be gradually implemented in all proofs.  

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_array_list_copy/aws_array_list_copy_harness.c [hunk 1]
```diff
@@ -16,61 +16,30 @@
 #include <aws/common/array_list.h>
 #include <proof_helpers/make_common_data_structures.h>
 
-/* These values allow us to reach a higher coverage rate */
-#define MAX_ITEM_SIZE 2
-#define MAX_INITIAL_ITEM_ALLOCATION (UINT64_MAX / MAX_ITEM_SIZE) + 1
-
 /**
- * Runtime: 0m13.098s
+ * Runtime: 32s
```
Comment (tautschnig) [inline .cbmc-batch/jobs/aws_array_list_copy/aws_array_list_copy_harness.c]: Is it the different proof style or a change in CBMC that has caused this runtime increase?

File: include/aws/common/array_list.inl [hunk 2]
```diff
@@ -77,12 +77,14 @@ void aws_array_list_init_static(
 
 AWS_STATIC_IMPL
 bool aws_array_list_is_valid(const struct aws_array_list *AWS_RESTRICT list) {
+    if(!list) {
+        return false;
+    }
     size_t required_size;
     bool required_size_is_valid = aws_mul_size_checked(list->length, list->item_size, &required_size) == AWS_OP_SUCCESS;
     bool current_size_is_valid = list->current_size >= required_size;
-    bool item_size_is_valid = list->item_size != 0;
```
Comment (tautschnig) [inline include/aws/common/array_list.inl]: Why is this being removed?

Comment (feliperodri) [reply include/aws/common/array_list.inl]: I tested the invariants with all functions and it turns out this particular one is not an invariant for the `aws_array_list` structure, only for the `aws_array_list_back` function. I'll include the pre- and post-conditions in all functions on a separate PR.

File: .cbmc-batch/jobs/aws_array_list_copy/aws_array_list_copy_harness.c (hunk 1 above)
Comment (feliperodri) [reply .cbmc-batch/jobs/aws_array_list_copy/aws_array_list_copy_harness.c]: The new improvements in CBMC actually lower the verification time to `10s`. I've updated it now.

**Pull request merged**